### PR TITLE
HTTP response generator

### LINKED ISSUE #14: Http builder
Generator for the http client requests

### PR DESCRIPTION
**Pull request by pmerku**
Proof of concept for a HTTP response builder. Feedback is appreciated :)

todo:
- [x] populate time for Date header
- [x] proper exception handling
- [x] make itoa not c++11
- [x] change example hardcoded values to the correct response values
- [x] change `AResponse::setResponse(...)` so it generates multiple types of responses

resolves #14

### CODE REVIEW

File: src/includes/server/http/response/AResponse.hpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		AResponse();
+		virtual ~AResponse();
+
+		const std::string &getResponse() const;
+		void setResponse(const NotApache::HTTPClient &client);
+	};
+
+} // namespace response
+
+#endif //ARESPONSE_HPP
```
Comment (mrjvs) [inline src/includes/server/http/response/AResponse.hpp]: why is this called AResponse? its not an abstract class.

File: src/includes/server/http/response/ResponseBuilder.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#include <cstring>
+#include <cerrno>
+
+#include <ctime>
+#include <sys/time.h> // have to use C header for gettimeofday()
+
+namespace response {
+
+	template<typename T, typename U>
+	class createMap {
```
Comment (mrjvs) [inline src/includes/server/http/response/ResponseBuilder.hpp]: use PascalCase for class names

File: src/includes/utils/intToString.hpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+//
+
+#ifndef INTTOSTRING_HPP
+#define INTTOSTRING_HPP
+
+#include <string>
+
+namespace utils {
+
+//	std::string intToString(int x);
```
Comment (mrjvs) [inline src/includes/utils/intToString.hpp]: ?

File: src/includes/utils/localTime.hpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
+//
+// Created by pmerku on 3/20/2021.
+//
+
+#ifndef LOCALTIME_HPP
+#define LOCALTIME_HPP
+
+#include <string>
```
Comment (mrjvs) [inline src/includes/utils/localTime.hpp]: include <ctime>

File: src/includes/utils/localTime.hpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace utils {
+
+	inline struct tm *relativeLocalTime(const time_t *t, struct tm *tm) {
+		(void)t;
+		return tm;
+	}
```
Comment (mrjvs) [inline src/includes/utils/localTime.hpp]: is this unfinished? what will it be doing?

File: src/includes/server/http/response/AResponse.hpp (hunk 1 above)
Comment (mrjvs) [reply src/includes/server/http/response/AResponse.hpp]: and it seems kinda pointless, the builder is supposed to be used directly in HTTPResponder::generateResponse()

File: src/utils/intToString.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	else
+		x *= -1;
+	str[len] = '\0';
+	while (len-- && x) {
+		str[len - 1] = -(x % 10) + '0';
+		x /= 10;
+	}
+	::free(str);
+	return std::string(str);
+}
```
Comment (mrjvs) [inline src/utils/intToString.cpp]: dont use malloc here, you can do all with std::string

File: src/server/http/response/ResponseBuilder.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		(403, "Forbidden");
+
+ResponseBuilder::ResponseBuilder(const std::string &protocol) {
+	this->_response = protocol;
+}
+
+ResponseBuilder &ResponseBuilder::setStatus(int code) {
+	std::map<int, std::string>::const_iterator it = _statusMap.find(code);
+	this->_response += " " + std::string("400"); //utils::intToString(it->first);
+	this->_response += " " + it->second;
```
Comment (mrjvs) [inline src/server/http/response/ResponseBuilder.cpp]: the builder should have all its parts (status code, headers, body, HTTP version) as private fields, then compile everything when you run build(). so for example, setStatus() would basically be `_status = code` with validation check to see if its a valid code.

Comment (mrjvs) [reply src/server/http/response/ResponseBuilder.cpp]: that would make it so the parts wouldnt have to be ran in a specific order

File: src/server/http/response/ResponseBuilder.cpp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	if (ret < 0)
+		throw std::runtime_error(strerror(errno)); // TODO exception handling
+	this->_response += std::string(date, ret) + "UTC";
+	return setEndLine();
+}
+
+ResponseBuilder &ResponseBuilder::setEndLine() {
+	this->_response += "\r\n";
+	return *this;
+}
```
Comment (mrjvs) [inline src/server/http/response/ResponseBuilder.cpp]: unnecessary, should be made on build() (check top comment of the file)

**Review by mrjvs — CHANGES_REQUESTED** [review]

File: src/includes/server/http/response/AResponse.hpp (hunk 1 above)
Comment (pmerku) [reply src/includes/server/http/response/AResponse.hpp]: I was thinking of making a general response class and then have a concrete class for HTTP and one for the Terminal. But right now yes it's pointless

File: src/includes/utils/localTime.hpp (hunk 5 above)
Comment (pmerku) [reply src/includes/utils/localTime.hpp]: Yes unfinished, this function will populate the time struct for strftime()

File: src/server/http/response/ResponseBuilder.cpp (hunk 7 above)
Comment (pmerku) [reply src/server/http/response/ResponseBuilder.cpp]: I will change this yes :)

File: src/includes/utils/localTime.hpp (hunk 4 above)
Comment (mrjvs) [reply src/includes/utils/localTime.hpp]: include ctime. markdown cut it out

**Review by mrjvs — APPROVED** [review]

**Review by maartenRamadan — APPROVED** [review]

**Pull request merged**